monitoring/node_collector.go

```diff
@@ -0,0 +1,77 @@
+package monitoring
+
+import (
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/types"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+)
+
+type NodeCollector struct {
+	*baseCollector
+	newNodeStatusMetric metricInfo
+}
+
+func NewNodeCollector(
+	logger logrus.FieldLogger,
+	nodeID string,
+	ds *datastore.DataStore) *NodeCollector {
+
+	nc := &NodeCollector{
+		baseCollector: newBaseCollector("node", logger, nodeID, ds),
+	}
+
+	nc.newNodeStatusMetric = metricInfo{
+		Desc: prometheus.NewDesc(
+			prometheus.BuildFQName(longhornName, subsystemNodeStats, "node_status"),
+			"Status of this node",
+			[]string{prometheusNodeIDLabel, prometheusConditionLabel},
+			nil,
+		),
+		Type: prometheus.GaugeValue,
+	}
+
+	return nc
+}
+
+func (nc *NodeCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- nc.newNodeStatusMetric.Desc
+}
+
+func (nc *NodeCollector) Collect(ch chan<- prometheus.Metric) {
+	nodeList, err := nc.ds.ListNodesRO()
+	if err != nil {
+		nc.logger.WithError(err).Warn("error during scrape")
+		return
+	}
+
+	nc.collectNodeStatus(ch, nodeList)
+
+}
+
+func (nc *NodeCollector) collectNodeStatus(ch chan<- prometheus.Metric, nodeList []*longhorn.Node) {
+	for _, node := range nodeList {
+		if node.Name != nc.currentNodeID {
+			continue
+		}
+
+		for _, conditionType := range []string{types.NodeConditionTypeReady, types.NodeConditionTypeSchedulable, types.NodeConditionTypeMountPropagation} {
+			val := 0
+			switch condition := types.GetCondition(node.Status.Conditions, conditionType); condition.Status {
+			case types.ConditionStatusTrue:
+				val = 1
+			case types.ConditionStatusFalse:
+				val = 2
+			case types.ConditionStatusUnknown:
+				val = 3
+			}
+			ch <- prometheus.MustNewConstMetric(nc.newNodeStatusMetric.Desc, nc.newNodeStatusMetric.Type, float64(val), nc.currentNodeID, strings.ToLower(conditionType))
```

Comment:
Add the condition_reason values as a dimension.
This allows for example to find how often a node became not ready because of network failure, cpu failure, etc.

Reply:
Thank you! Fixed